dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/Json.java

```diff
@@ -0,0 +1,136 @@
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.io.IOException;
+import java.io.InputStream;
+import javax.annotation.Nullable;
+
+public final class Json implements DeserializerFactory, SerializerFactory {
+
+    private static final BodySerDe DEFAULT_BODY_SERDE =
+            DefaultConjureRuntime.builder().encodings(Encodings.json()).build().bodySerDe();
+
+    private final BodySerDe bodySerDe;
+
+    public Json() {
+        this(DEFAULT_BODY_SERDE);
+    }
+
+    public Json(ObjectMapper mapper) {
+        this(DefaultConjureRuntime.builder().encodings(json(mapper)).build().bodySerDe());
+    }
+
+    private Json(BodySerDe bodySerDe) {
+        this.bodySerDe = bodySerDe;
+    }
+
+    @Override
+    public <T> com.palantir.dialogue.Deserializer<T> deserializerFor(TypeMarker<T> type) {
+        return bodySerDe.deserializer(type);
+    }
+
+    @Override
+    public <T> com.palantir.dialogue.Serializer<T> serializerFor(TypeMarker<T> type) {
+        return bodySerDe.serializer(type);
+    }
+
+    // Copied code, tbd how to share.
+
+    private static Encoding json(ObjectMapper mapper) {
+        return new AbstractJacksonEncoding(mapper) {
+            private static final String CONTENT_TYPE = "application/json";
+
+            @Override
+            public String getContentType() {
+                return CONTENT_TYPE;
+            }
+
+            @Override
+            public boolean supportsContentType(String contentType) {
+                return matchesContentType(CONTENT_TYPE, contentType);
+            }
+        };
+    }
+
+    private abstract static class AbstractJacksonEncoding implements Encoding {
+
+        private final ObjectMapper mapper;
+
+        AbstractJacksonEncoding(ObjectMapper mapper) {
+            this.mapper = Preconditions.checkNotNull(mapper, "ObjectMapper is required");
+        }
+
+        @Override
+        public final <T> Serializer<T> serializer(TypeMarker<T> type) {
+            ObjectWriter writer = mapper.writerFor(mapper.constructType(type.getType()));
+            return (value, output) -> {
+                Preconditions.checkNotNull(value, "cannot serialize null value");
+                try {
+                    writer.writeValue(output, value);
+                } catch (IOException e) {
+                    throw new SafeRuntimeException("Failed to serialize payload", e);
+                }
+            };
+        }
+
+        @Override
+        public final <T> Deserializer<T> deserializer(TypeMarker<T> type) {
+            ObjectReader reader = mapper.readerFor(mapper.constructType(type.getType()));
+            return input -> {
+                try (InputStream inputStream = input) {
+                    T value = reader.readValue(inputStream);
+                    // Bad input should result in a 4XX response status, throw IAE rather than NPE.
```

Comment:
I expect we copied this from conjure-undertow, Dialogue should cause a 5xx if this happens. We can fix the core dialoge implementation separately.

Reply:
That fix: #1148

Reply:
Alright, I'll recopy the code.

Reply:
Done